### PR TITLE
Admin Menu

### PR DESCRIPTION
**Pull request by beaar**
Don't try this at home! 🔥 ⚠️ 
The current state is little more than a proof of concept. There's a lot of dumb/ugly/unfinished/obsolete code. Don't look 😩

The default keybind is <kbd>Shift</kbd><kbd>F1</kbd>, but it uses CBA keybinds so you can easily change it. There's an array in root config (see config.cpp) for authorized player uids.

Screenshot

**Current state of things and stuff**
- Dashboard is pretty much done, is missing a list of persons with Zeus access. Waiting for toggleable safe start (#144).
- Player Management tab has received the most attention. Player selection is mostly working, with a few implemented "utilities" to try it out. Waiting for extra spectator data (#144).
- The ending tab needs back-end work.
- Respawn and logging tabs aren't started on.
- I've written the system time extension (#144) for the log tab but will be committing it separately.
- The entire thing could benefit from some refactoring.

### CODE REVIEW

File: addons/adminmenu/XEH_postInit.sqf [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#include "\x\tmf\addons\adminmenu\script_component.hpp"
+
+#include "initKeybinds.sqf"
+
+GVAR(sideColors) = [
+	[profilenamespace getvariable ['Map_BLUFOR_R',0], profilenamespace getvariable ['Map_BLUFOR_G',0], profilenamespace getvariable ['Map_BLUFOR_B',1], 0.8],
```
Comment (headswe) [inline addons/adminmenu/XEH_postInit.sqf]: use TMF_common_fnc_sideToColor

File: addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	TMF_ADMINMENU_PMAN_X_SIDE,
+	TMF_ADMINMENU_PMAN_X_QRESP,
+	TMF_ADMINMENU_PMAN_X_STEAM
+];
+private _stdWidth = TMF_ADMINMENU_STD_WIDTH;
+private _stdHeight = 0.978 * TMF_ADMINMENU_STD_HEIGHT;
+private _ctrlHeight = ((ctrlPosition _list) param [3]) / ((lbSize _list) max 1);
+
+diag_log "_stdHeight _ctrlHeight a/b";
+diag_log _stdHeight;
```
Comment (headswe) [inline addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf]: debug lines

File: addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+diag_log _ctrlHeight;
+diag_log (_stdHeight / _ctrlHeight);
```
Comment (headswe) [inline addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf]: Debug lines

File: addons/adminmenu/functions/fn_modal_steamprofile.sqf [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		_text = format [
+			"%1<t size='1' align='right'><a color='#FFC04D' href='http://steamcommunity.com/profiles/%2'>%3</a></t><br/>", 
+			_text,
+			getPlayerUID _x,
+			name _x
+		];
+	};
+} forEach _data;
+
+diag_log _text;
```
Comment (headswe) [inline addons/adminmenu/functions/fn_modal_steamprofile.sqf]: Debug lines

File: addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+params ["_display"];
+
+private _list = _display displayCtrl IDC_TMF_ADMINMENU_PMAN_LIST;
+private _ctrlGrp = ctrlParentControlsGroup _list;
+
+private _numControls = count GVAR(playerManagement_listControls);
+private _adjustControls = (lbSize _list) - _numControls;
+
+//systemChat format ["CONTROL PRE  Players: %1 | List rows: %2 | _adjustControls: %3 | Controls: %4", count GVAR(playerManagement_players), lbSize _list, _adjustControls, count GVAR(playerManagement_listControls)];
+diag_log format ["CONTROL PRE  Players: %1 | List rows: %2 | _adjustControls: %3 | Controls: %4", count GVAR(playerManagement_players), lbSize _list, _adjustControls, count GVAR(playerManagement_listControls)];
```
Comment (headswe) [inline addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf]: Debug lines

File: addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		"<a href='http://steamcommunity.com/profiles/%1' size='0.9'><img image='%2'/></a>", 
+		name _player,
+		"\a3\ui_f\data\GUI\RscCommon\RscButtonMenuSteam\steam_ca.paa"
+	];
+	//_ctrlSteam ctrlSetStructuredText parseText format ["<a href='http://steamcommunity.com/profiles/%1' size='0.8'>S</a>", getPlayerUID _player];
+	//_ctrlSteam ctrlSetStructuredText parseText format ["<a href='http://steamcommunity.com/profiles/%1' size='0.8'>S</a>", name _player];
+} forEach GVAR(playerManagement_listControls);
+
+//systemChat format ["CONTROL POST Players: %1 | List rows: %2 | _adjustControls: %3 | Controls: %4", count GVAR(playerManagement_players), lbSize _list, _adjustControls, count GVAR(playerManagement_listControls)];
+diag_log format ["CONTROL POST Players: %1 | List rows: %2 | _adjustControls: %3 | Controls: %4", count GVAR(playerManagement_players), lbSize _list, _adjustControls, count GVAR(playerManagement_listControls)];
```
Comment (headswe) [inline addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf]: Debug lines

File: addons/adminmenu/functions/fn_dashboard.sqf [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	(_display displayCtrl IDC_TMF_ADMINMENU_DASH_HEADLESS) ctrlSetText str (count entities "HeadlessClient_F");
+	(_display displayCtrl IDC_TMF_ADMINMENU_DASH_VIRTUALS) ctrlSetText str (count entities "VirtualCurator_F");
+
+	//private _deadUnits = ((allPlayers - _liveUnits) - _spectators) - _headlessClients;
+
+	private _liveUnits = allUnits;
+	private _spectatorUnits = (entities QEGVAR(spectator,unit)) select {isPlayer _x};
+	{
+		_x params ["_ai", "_players", "_spectators", "_total"];
+		private _side = [blufor, opfor, resistance, civilian] param [_forEachIndex];
```
Comment (cptnnick) [inline addons/adminmenu/functions/fn_dashboard.sqf]: `select` is better here

File: addons/adminmenu/functions/fn_endMissionCommit.sqf [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	};
+
+	// Per-side endings
+	case IDC_TMF_ADMINMENU_ENDM_SIDESPECIFIC: {
+		private _isDraw = cbChecked (_display displayCtrl IDC_TMF_ADMINMENU_ENDM_SIDEDRAW);
+		if (_isDraw) then {
+			"RespawnRoundsTie" remoteExec [_endMissionFunc];
+		} else {
+			private _winning = [];
+			private _sides = [blufor, opfor, resistance, civilian];
```
Comment (cptnnick) [inline addons/adminmenu/functions/fn_endMissionCommit.sqf]: this doesn't need a variable, only used once

File: addons/adminmenu/functions/fn_endMissionCommit.sqf [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+					_winning pushBack (_sides select _forEachIndex);
+				};
+			} forEach [QGVAR(endingSideBlufor), QGVAR(endingSideOpfor), QGVAR(endingSideIndependent), QGVAR(endingSideCivilian)];
+
+			[[_endMissionFunc, _winning], {
+				//private _ending = ["LOOSER", "WINNER"] select (playerSide in (param [1])); // looser is BI's typo not mine ok
+				//_ending call (param [0]);
+				private _isVictory = playerSide in (param [1]);
+				["", _isVictory] call (param [0]); // win/lose auto from _isVictory ?
+			}] remoteExec ["BIS_fnc_call", -2];
```
Comment (cptnnick) [inline addons/adminmenu/functions/fn_endMissionCommit.sqf]: This is non-ideal, why not just make this a function?

File: addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+private _xPos = [
+	TMF_ADMINMENU_PMAN_X_DEAD,
+	TMF_ADMINMENU_PMAN_X_SIDE,
+	TMF_ADMINMENU_PMAN_X_QRESP,
+	TMF_ADMINMENU_PMAN_X_STEAM
+];
+private _stdWidth = TMF_ADMINMENU_STD_WIDTH;
+private _stdHeight = 0.978 * TMF_ADMINMENU_STD_HEIGHT;
+private _ctrlHeight = ((ctrlPosition _list) param [3]) / ((lbSize _list) max 1);
```
Comment (cptnnick) [inline addons/adminmenu/functions/fn_playerManagementUpdateControls.sqf]: Another param

File: addons/adminmenu/functions/fn_utility_assigngear.sqf [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+private _ctrlCheckFromMission = _display ctrlCreate ["RscCheckBox", -1, _ctrlGroup];
+GVAR(utilityTabControls) pushBack _ctrlCheckFromMission;
+_ctrlCheckFromMission ctrlSetPosition [(2 * TMF_ADMINMENU_STD_WIDTH) + (0.25 * _ctrlGrpWidth), 0, TMF_ADMINMENU_STD_WIDTH, TMF_ADMINMENU_STD_HEIGHT];
+_ctrlCheckFromMission ctrlCommit 0;
+_ctrlCheckFromMission ctrlAddEventHandler ["onCheckedChanged", {
+    params ["_ctrlCheckBox", "_onlyPresent"];
+    GVAR(utility_assigngear_onlypresent) = [false, true] select _onlyPresent;
+
+    _factionsFound = [];
```
Comment (cptnnick) [inline addons/adminmenu/functions/fn_utility_assigngear.sqf]: Seems no longer used?

File: addons/adminmenu/functions/fn_utility_assigngear.sqf [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+            _playerRole = "r";
+            _tickCheckbox = true;
+        } else {
+            private _curSelRole = lbCurSel _ctrlComboRole;
+            if !(_curSelRole isEqualTo -1) then {
+                if (_ctrlComboRole lbData _curSelRole )
+            };
+        };
+
+        while {(count _roles) < (lbSize _ctrlComboRole)} do {
```
Comment (cptnnick) [inline addons/adminmenu/functions/fn_utility_assigngear.sqf]:
```sqf"
for "_i" from 0 to ((lbSize _ctrlComboRole) - (count _roles)) do
...
```
so it doesnt have to evaluate the condition every iteration.

**Pull request merged**